coderd/database/dbmem/dbmem.go

```diff
@@ -6083,6 +6083,47 @@ func (q *FakeQuerier) RegisterWorkspaceProxy(_ context.Context, arg database.Reg
 	return database.WorkspaceProxy{}, sql.ErrNoRows
 }
 
+func (q *FakeQuerier) RestrictWorkspaceAgentPortSharesByTemplate(_ context.Context, arg database.RestrictWorkspaceAgentPortSharesByTemplateParams) error {
+	err := validateDatabaseType(arg)
+	if err != nil {
+		return err
+	}
+
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	var sharesToDelete []database.WorkspaceAgentPortShare
```

Comment:
You could store the a list of each index to delete and then iterate over it in reverse order